### PR TITLE
feat(license): Add a script to convert license headers in some files …

### PR DESCRIPTION
**Pull request by nickvergessen**
…in a programmatic way

cc @AndyScherzinger

### CODE REVIEW

**Review by AndyScherzinger — APPROVED** [review]

File: spdx-convertor/convert.php [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+$isDryRun = $argv[1] === '--dry-run';
+$path = $isDryRun ? $argv[2] : $argv[1];
+
+function generateSpdxContent(string $originalHeader, string $file): array {
+	$newHeaderLines = [];
+	$authors = [];
+	$license = null;
+
+	foreach (explode("\n", $originalHeader) as $line) {
+		// @copyright Copyright (c) 2023 Joas Schilling <coding@schilljs.com>
```
Comment (Antreesy) [inline spdx-convertor/convert.php:24]: There's a minor issue with creation years in copyright headers:

- If several copyrights were in the file before, that leads to duplicated headers:
  ```
  /**
   * SPDX-FileCopyrightText: 2023 Nextcloud GmbH and Nextcloud contributors
   * SPDX-FileCopyrightText: 2022 Nextcloud GmbH and Nextcloud contributors
   * SPDX-License-Identifier: AGPL-3.0-or-later
   */
   ```
  We should keep only the earliest year in that case
- If copyright doesn't have a year listed, that leads to non-specified header:
  ```
  /**
   * SPDX-FileCopyrightText: Nextcloud GmbH and Nextcloud contributors
   * SPDX-License-Identifier: AGPL-3.0-or-later
   */
   ```
  We may extract the creation year from the first known commit to that file (shell script example):
  `creation_year=$(git log --diff-filter=A --follow --format=%aI "$file" | tail -n 1 | cut -d'-' -f1)`

Comment (nickvergessen) [reply spdx-convertor/convert.php:24]: The ones without a year is where I added it manually

Comment (Antreesy) [reply spdx-convertor/convert.php:24]: Not sure, if that would be complex to automate for files without license as well

**Review by AndyScherzinger — APPROVED** [review]

File: spdx-convertor/convert.php (hunk 1 above)
Comment (nickvergessen) [reply spdx-convertor/convert.php:24]: I think its better to have a manual look when its missing.
But the duplicate one I will solve

File: spdx-convertor/convert.php [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+ */
+
+require_once __DIR__ . '/vendor/autoload.php';
+
+if ($argc < 2 || in_array('--help', $argv) || in_array('-h', $argv)) {
+	die("convert.php [--dry-run] <path>\n");
+}
+
+$isDryRun = $argv[1] === '--dry-run';
+$path = $isDryRun ? $argv[2] : $argv[1];
```
Comment (kesselb) [inline spdx-convertor/convert.php:16]: Lazy me called the script without a trailing slash (`php convert.php ../../nextcloud/server/apps-extra/user_usage_report`)

and ended up with a `user_usage_reportAUTHORS.md` in my apps-extra folder :roll_eyes:

File: spdx-convertor/convert.php (hunk 1 above)
Comment (nickvergessen) [reply spdx-convertor/convert.php:25]: Done

File: spdx-convertor/convert.php (hunk 2 above)
Comment (nickvergessen) [reply spdx-convertor/convert.php]: Fixed

**Comment by nickvergessen** [comment]
I'll check it, but that sounds not right... but might be a cause of `/` vs `\`

**Comment by nickvergessen** [comment]
Clarified a bit the error cases:
image


Should be good to go now.

**Comment by AndyScherzinger** [comment]
👍 works fine for me now (on Windows) 👍

**Comment by AndyScherzinger** [comment]
@kesselb fancy to give it another try? Works fine for me now

**Comment by kesselb** [comment]
```
php convert.php --dry-run ../../nextcloud/server/apps-extra/user_usage_report/ 
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/CHANGELOG.md skipped
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/README.md skipped
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/appinfo/info.xml skipped
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/composer.json skipped
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/composer.lock skipped
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/docs/screenshot.png skipped
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/AppInfo/Application.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Command/Generate.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Formatter.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Listener.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Migration/Version1001Date20180806133516.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Migration/Version1004Date20200626143122.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Migration/Version1004Date20200626143419.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Migration/Version1004Date20200630071803.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Reports/AllUsers.php would replace
 ├─ ☑️  /code/nextcloud/server/apps-extra/user_usage_report/lib/Reports/SingleUser.php would replace
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/psalm.xml skipped
 ├─ ◽ /code/nextcloud/server/apps-extra/user_usage_report/tests/psalm-baseline.xml skipped
 ├─ ❌ /code/nextcloud/server/apps-extra/user_usage_report/COPYING Not handled
 ├─ ❌ /code/nextcloud/server/apps-extra/user_usage_report/Makefile Not handled
 └─ ✅ Can generate AUTHORS.md

# Authors

- Joas Schilling <coding@schilljs.com>
```

Looks good, Joas is properly listed as author.

**Pull request merged**